src/olmo_core/train/callbacks/batch_size_scheduler.py

```diff
@@ -0,0 +1,171 @@
+        # the "initial_lr_field" in the optimizer state always gets reset to the value from
+        # the config, not the checkpoint, and same for the "batches_processed" field in the
+        # data loader.
+        self._maybe_update_batch_size_and_lr()
+
+    def pre_load_batch(self):
+        self._maybe_update_batch_size_and_lr()
+
+    def post_step(self):
+        self.trainer.record_metric("train/global batch size", self.current_batch_size)
+
+    def _maybe_update_batch_size_and_lr(self):
+        # Find latest event in the schedule to apply.
+        for target_batch_size, event_start in reversed(list(zip(self.batch_sizes, self.schedule))):
+            if event_start.due(
+                step=self.trainer.global_step,
+                tokens=self.trainer.global_train_tokens_seen,
+                epoch=self.trainer.epoch,
+            ):
+                if target_batch_size > self.current_batch_size:
+                    assert target_batch_size % self.current_batch_size == 0
+                    # When increasing the batch size, need to wait until 'batches_processed' is divisible
+                    # by the factor that batch size is being increased by.
+                    ratio = target_batch_size // self.current_batch_size
+                    if self.trainer.data_loader.batches_processed % ratio == 0:
+                        self._update_batch_size_and_lr(target_batch_size)
+                elif target_batch_size < self.current_batch_size:
+                    assert self.current_batch_size % target_batch_size == 0
+                    self._update_batch_size_and_lr(target_batch_size)
+
+                break
+
+    def _update_batch_size_and_lr(self, batch_size: int):
+        log.info(f"Changing global batch size to {batch_size:,d}...")
+        ratio = batch_size / self.current_batch_size
+        lr_adjustment_factor = math.sqrt(ratio)
+        self.trainer.data_loader.global_batch_size = batch_size
+
+        optimizers: Optional[List[torch.optim.Optimizer]] = None
+        scheduler: Optional[Scheduler] = None
+        if isinstance(self.trainer.train_module, TransformerTrainModule):
+            optimizers = [self.trainer.train_module.optim]
+            scheduler = self.trainer.train_module.scheduler
+        elif isinstance(self.trainer.train_module, TransformerPipelineTrainModule):
+            optimizers = self.trainer.train_module.optimizers
+            scheduler = self.trainer.train_module.scheduler
+
+        if not optimizers:
+            log.warning(
+                f"Unable to adjust learning rate for {self.trainer.train_module.__class__.__name__} train module class"
+            )
+            return
+
+        for optim in optimizers:
+            if not isinstance(optim, (torch.optim.Adam, torch.optim.AdamW, SkipStepAdamW)):
+                log.warning(
+                    f"Unable to adjust learning rate for {optim.__class__.__name__} optimizer"
+                )
+                return
```

Comment:
It falls back to doing nothing? I'd rather it fails. Otherwise we end up running 500B tokens worth of baselines, only to discover afterwards that something we configured didn't actually run.

Reply:
done: ff895e8f